src/const_permutation.rs

```diff
@@ -0,0 +1,440 @@
+        )
+    }
+
+    /// Returns a random permutation.
+    ///
+    /// Uses a uniform distribution.
+    #[cfg(feature = "random")]
+    pub fn random<R>(rng: &mut R) -> Self
+    where
+        R: Rng,
+    {
+        let ps = Permutations::new(N);
+        let i = rng.gen_range(0..ps.len());
+        ps.get(i).expect("random index out of range")
+    }
+    /// Applies the permutation to an element.
+    pub fn apply(&self, i: usize) -> usize {
+        self.0[i]
+    }
+    /// Returns an array permuted by this permutation.
+    pub fn permute<T>(&self, a: &[T; N]) -> [T; N]
+    where
+        T: Clone + Debug,
+    {
+        (0..N)
+            .map(|i| a[self.apply(i)].clone())
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap()
+    }
+    /// Returns the composition of the permutation with itself.
+    pub fn square(&self) -> Self {
+        self * self
+    }
+    /// Returns the composition of the permutation with itself `exp` number of times.
+    pub fn pow(&self, exp: u32) -> Self {
+        if exp == 0 {
+            ConstPermutation::identity()
+        } else if exp == 1 {
+            self.clone()
+        } else if exp % 2 == 0 {
+            self.square().pow(exp / 2)
+        } else {
+            self * self.pow(exp - 1)
+        }
+    }
+    /// Returns the inverse permutation.
+    pub fn inv(&self) -> Self {
+        let mut map = [0; N];
+        for i in 0..N {
+            let j = self.0[i];
+            map[j] = i;
+        }
+        ConstPermutation(map)
+    }
+    /// The length of the permutation.
+    ///
+    /// That is, the length of its domain and range.
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+    /// Returns true if this is the empty permutation.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+    /// Returns true if the permutation is even.
+    pub fn is_even(&self) -> bool {
+        let mut even = true;
+        let mut seen = vec![false; self.len()];
+        for i in 0..self.len() {
+            if !seen[i] {
+                seen[i] = true;
+                let mut j = self.apply(i);
+                while j != i {
+                    seen[j] = true;
+                    j = self.apply(j);
+                    even = !even;
+                }
+            }
+        }
+        even
+    }
+    /// Returns true if the permutation is odd.
+    pub fn is_odd(&self) -> bool {
+        !self.is_even()
+    }
+    /// Returns the sign of a permutation.
+    pub fn sign(&self) -> isize {
+        if self.is_even() {
+            1
+        } else {
+            -1
+        }
+    }
+}
+impl<const N: usize> ops::Mul<ConstPermutation<N>> for ConstPermutation<N> {
+    type Output = ConstPermutation<N>;
+    fn mul(self, other: ConstPermutation<N>) -> Self::Output {
+        let mut map = [0; N];
+        for i in 0..N {
+            map[i] = other.apply(self.apply(i));
+        }
+        ConstPermutation(map)
```

Comment:
Is this just `permute`?

```suggestion
        ConstPermutation(other.permute(&self.0))
```